src/Gui/Action.cpp

```diff
@@ -1243,4 +1350,206 @@ void WindowAction::addTo ( QWidget * w )
+    {
+        return this->createIndex(row, 0);
+    }
+
+    virtual int rowCount(const QModelIndex &) const
+    {
+        return (int)(_Commands.size());
+    }
+
+    virtual int columnCount(const QModelIndex &) const
+    {
+        return 1;
+    }
+};
+
+
+// --------------------------------------------------------------------
+
+CommandCompleter::CommandCompleter(QLineEdit *lineedit, QObject *parent)
+    : QCompleter(parent)
+{
+    this->setModel(new CommandModel(this));
+#if QT_VERSION>=QT_VERSION_CHECK(5,2,0)
+    this->setFilterMode(Qt::MatchContains);
+#endif
+    this->setCaseSensitivity(Qt::CaseInsensitive);
+    this->setCompletionMode(QCompleter::PopupCompletion);
+    this->setWidget(lineedit);
+    connect(lineedit, SIGNAL(textEdited(QString)), this, SLOT(onTextChanged(QString)));
+    connect(this, SIGNAL(activated(QModelIndex)), this, SLOT(onCommandActivated(QModelIndex)));
+    connect(this, SIGNAL(highlighted(QString)), lineedit, SLOT(setText(QString)));
+}
+
+bool CommandCompleter::eventFilter(QObject *o, QEvent *ev)
+{
+    if (ev->type() == QEvent::KeyPress
+            && (o == this->widget() || o == this->popup()))
+    {
+        QKeyEvent * ke = static_cast<QKeyEvent*>(ev);
+        switch(ke->key()) {
+        case Qt::Key_Escape: {
+            auto edit = qobject_cast<QLineEdit*>(this->widget());
+            if (edit && edit->text().size()) {
+                edit->setText(QString());
+                popup()->hide();
+                return true;
+            } else if (popup()->isVisible()) {
+                popup()->hide();
+                return true;
+            }
+            break;
+        }
+        case Qt::Key_Tab: {
+            if (this->popup()->isVisible()) {
+                QKeyEvent kevent(ke->type(),Qt::Key_Down,0);
+                qApp->sendEvent(this->popup(), &kevent);
+                return true;
+            }
+            break;
+        }
+        case Qt::Key_Backtab: {
+            if (this->popup()->isVisible()) {
+                QKeyEvent kevent(ke->type(),Qt::Key_Up,0);
+                qApp->sendEvent(this->popup(), &kevent);
+                return true;
+            }
+            break;
+        }
+        case Qt::Key_Enter:
+        case Qt::Key_Return:
+            if (o == this->widget()) {
+                auto index = currentIndex();
+                if (index.isValid())
+                    onCommandActivated(index);
+                else
+                    complete();
+                ev->setAccepted(true);
+                return true;
+            }
+        default:
+            break;
+        }
+    }
+    return QCompleter::eventFilter(o, ev);
+}
+
+void CommandCompleter::onCommandActivated(const QModelIndex &index)
+{
+    QByteArray name = completionModel()->data(index, Qt::UserRole).toByteArray();
+    Q_EMIT commandActivated(name);
+}
+
+void CommandCompleter::onTextChanged(const QString &txt)
+{
+    if (txt.size() < 3 || !widget())

```

Comment:
The 3 needs some documentation here, otherwise it looks like a magic number.

Reply:
Comments added. To improve response time by not activate completer if text is too short.